poly/promise/promise.go

```diff
@@ -0,0 +1,744 @@
+ *   result. In short, make sure to verify only shares that are allotted to you.
+ */
+func (p *Promise) VerifyShare(i int, gKeyPair *config.KeyPair) bool {
+	if !p.validShare(i, gKeyPair) {
+		return false
+	}
+	diffieBase := p.shareSuite.Point().Mul(p.pubKey, gKeyPair.Secret)
+	share := p.diffieHellmanDecrypt(p.secrets[i], diffieBase)
+	return p.pubPoly.Check(i, share)
+}
+
+/* Verifies whether a Promise objects has enough signatures to be valid.
+ *
+ * Return
+ *   whether the Promise is now valid and considered trustworthy.
+ *
+ * Technical Notes: The function goes through the list of signatures and checks
+ *                  whether the signature is properly signed. If at least r of
+ *                  these are signed and r is greater than t (the minimum number
+ *                  of shares needed to reconstruct the secret), the promise is
+ *                  considered valid.
+ */
+func (p *Promise) VerifyPromise() bool {
+	validSigs := 0
+	for i := 0; i < p.n; i++ {
+		// Check whether the signature is initialized. Otherwise, bad
+		// things will happen.
+		if p.signatures[i] != nil && p.VerifySignature(p.signatures[i]) {
+			validSigs += 1
+		}
+	}
+	return p.r > p.t && validSigs >= p.r
+}
+
+/* Produce a signature for a given guardian
+ *
+ * Arguments
+ *    i         = the index of the guardian's share
+ *    gKeyPair  = the public/private keypair of the guardian.
+ *
+ * Return
+ *   A PromiseSignature object with the signature.
+ *
+ * Note:
+ *   The signature message will always be of the form:
+ *      Guardian approves PromiseId
+ *
+ *   It is assumed that the guardian has called VerifyShare first and hence
+ *   it is assumed that the input to the function is trusted.
+ */
+func (p *Promise) Sign(i int, gKeyPair *config.KeyPair) *PromiseSignature {
+	set        := anon.Set{gKeyPair.Public}
+	approveMsg := gKeyPair.Public.String() + " approves " + p.id
+	sig        := anon.Sign(gKeyPair.Suite, random.Stream, []byte(approveMsg),
+		set, nil, 0, gKeyPair.Secret)
+		
+	return new(PromiseSignature).Init(i, gKeyPair.Suite, sig)
+}
+
+/* An internal helper function, makes sure that a promise signature is formatted
+ * properly and that no data was sent maliciously (index out of bounds, etc.)
+ *
+ * Arguments
+ *    sig = the PromiseSignature to check
+ *
+ * Return
+ *   whether the PromiseSignature was formatted properly
+ *
+ * Note:
+ *   Please see VerifySignature for more on validating signatures.
+ */
+func (p *Promise) validSignature(sig *PromiseSignature) bool {
+	if sig.pi < 0 || sig.pi > p.n {
+		return false
+	}
+	
+	if sig.signature == nil {
+		return false
+	}
+	
+	return true
+}
+
+/* Verifies a signature from a given guardian
+ *
+ * Arguments
+ *    sig = the PromiseSignature object containing the signature
+ *
+ * Return
+ *   whether or not the signature is valid
+ */
+func (p *Promise) VerifySignature(sig *PromiseSignature) bool {
+	if !p.validSignature(sig) {
```

Comment:
Made the changes. I was initially thinking about combining the two functions but hadn't gotten around to it quite yet.